fedimint-core/src/query.rs

```diff
@@ -252,6 +256,277 @@ impl<R: Eq + Clone + Debug> QueryStrategy<R> for CurrentConsensus<R> {
+                    .map(|peer_version| client_api_version.minor <= peer_version.minor)
+                    .unwrap_or(false)
+            })
+            .count();
+
+        if best_major_peer_num < peers_compatible_num {
+            best_major = Some(client_api_version);
+            best_major_peer_num = peers_compatible_num;
+        }
+    }
+
+    best_major
+}
+
+#[test]
+fn discover_common_core_api_version_sanity() {
+    use fedimint_core::module::MultiApiVersion;
+
+    let core_consensus = 0.into();
+    let client_versions = SupportedCoreApiVersions {
+        core_consensus,
+        api: MultiApiVersion::try_from_iter([
+            ApiVersion { major: 2, minor: 3 },
+            ApiVersion { major: 3, minor: 1 },
+        ])
+        .unwrap(),
+    };
+
+    assert!(discover_common_core_api_version(&client_versions, BTreeMap::from([])).is_none());
+    assert_eq!(
+        discover_common_core_api_version(
+            &client_versions,
+            BTreeMap::from([(
+                PeerId(0),
+                SupportedCoreApiVersions {
+                    core_consensus: 0.into(),
+                    api: MultiApiVersion::try_from_iter([ApiVersion { major: 2, minor: 4 }])
+                        .unwrap(),
+                }
+            )])
+        ),
+        Some(ApiVersion { major: 2, minor: 3 })
+    );
+    assert_eq!(
+        discover_common_core_api_version(
+            &client_versions,
+            BTreeMap::from([(
+                PeerId(0),
+                SupportedCoreApiVersions {
+                    core_consensus: 1.into(), // wrong consensus version
+                    api: MultiApiVersion::try_from_iter([ApiVersion { major: 2, minor: 4 }])
+                        .unwrap(),
+                }
+            )])
+        ),
+        None
+    );
+    assert_eq!(
+        discover_common_core_api_version(
+            &client_versions,
+            BTreeMap::from([
+                (
+                    PeerId(0),
+                    SupportedCoreApiVersions {
+                        core_consensus,
+                        api: MultiApiVersion::try_from_iter([ApiVersion { major: 2, minor: 2 }])
+                            .unwrap(),
+                    }
+                ),
+                (
+                    PeerId(1),
+                    SupportedCoreApiVersions {
+                        core_consensus,
+                        api: MultiApiVersion::try_from_iter([ApiVersion { major: 2, minor: 1 }])
+                            .unwrap(),
+                    }
+                ),
+                (
+                    PeerId(1),
+                    SupportedCoreApiVersions {
+                        core_consensus,
+                        api: MultiApiVersion::try_from_iter([ApiVersion { major: 3, minor: 1 }])
+                            .unwrap(),
+                    }
+                )
+            ])
+        ),
+        Some(ApiVersion { major: 3, minor: 1 })
```

Comment:
This is a bit confusing, shouldn't we return the major version which is the most supported by the peers, `2` in this case?

Reply:
A 2.x server would need to at least support minor version 3, so the 2.x servers present here are all out. I hope we'll never have to deal with that grade of incompatibility in one federation though, that would weaken censorship resistance since we essentially only talk to 1 guardian server in this case here.